quint/src/parsing/quintParserFrontend.ts

```diff
@@ -233,8 +213,79 @@ export function parsePhase2sourceResolution(
   for (const mods of sourceToModules.values()) {
     allModules = allModules.concat(mods)
   }
-  allModules.sort((m1, m2) => moduleRank.get(m2.name)! - moduleRank.get(m1.name)!)
-  return { ...mainPhase1Result, modules: allModules, sourceMap }
+  // sort the modules
+  const sortingResult = sortModules(allModules)
+
+  return {
+    ...mainPhase1Result,
+    errors: mainPhase1Result.errors.concat(sortingResult.errors),
+    modules: sortingResult.modules,
+    sourceMap,
+  }
+}
+
+/**
+ * Sort modules according to their imports, that is, importees should appear before importers.
+ * @param modules the modules to sort
+ * @return a structure that contains errors (if any were found) and the modules (sorted if no errors)
+ */
+function sortModules(modules: QuintModule[]): { errors: QuintError[]; modules: QuintModule[] } {
+  const idToModule = modules.reduce((accuMap, mod) => accuMap.set(mod.id, mod), ImmutMap<bigint, QuintModule>())
+  const nameToModule = modules.reduce((accuMap, mod) => accuMap.set(mod.name, mod), ImmutMap<string, QuintModule>())
+  if (nameToModule.size < modules.length) {
+    // sort the modules by their names
+    modules.sort((m1, m2) => m1.name.localeCompare(m2.name))
+    // find the first two adjacent modules that have the same name
+    let prev: Maybe<QuintModule> = none()
+    for (const mod of modules) {
+      if (prev.isJust() && prev.value.name === mod.name) {
+        const err: QuintError = {
+          code: 'QNT101',
+          message: `Two modules have a conflict on the same name: ${mod.name}`,
+          reference: mod.id,
+        }
+        return { errors: [err], modules }
+      }
+      prev = just(mod)
+    }
+    // this line should not be reachable, unless there is a bug above
+    throw new Error('This line should not be reached. Report a bug.')
+  }
+  // create the import graph
+  let edges = ImmutMap<bigint, ImmutSet<bigint>>()
+  for (const mod of modules) {
+    let imports = ImmutSet<bigint>()
+    for (const decl of mod.declarations) {
+      if (decl.kind === 'import' || decl.kind === 'instance') {
```

Comment:
Should this also sort exports?

```suggestion
      if (decl.kind === 'import' || decl.kind === 'instance' || decl.kind === 'export') {
```

Reply:
@bugarela I have accepted your suggestion, but then the failing tests showed me that we should not keep track of exports. I have added a clarifying comment. It's good that we have tests!